src/Manager.php

```diff
@@ -51,24 +51,62 @@ public function userHasPermission(
             return false;
         }
 
-        if ($userId === null) {
-            return $this->guestHasPermission($permission, $parameters);
-        }
+        $items = array_merge(
+            [$permission->getName() => $permission],
+            $this->itemsStorage->getParents($permission->getName()),
+        );
 
-        $permission = $this->itemsStorage->getPermission($permissionName);
-        if ($permission === null) {
-            return false;
+        if ($userId !== null) {
+            $userId = (string) $userId;
+            $userItemNames = $this->assignmentsStorage->filterUserItemNames($userId, array_keys($items));
+
+            $userItems = [];
+            foreach ($userItemNames as $itemName) {
+                $userItems[$itemName] = $items[$itemName];
+            }
+
+            foreach ($userItems as $userItem) {
+                foreach ($this->itemsStorage->getAllChildren($userItem->getName()) as $childItem) {
+                    if (array_key_exists($childItem->getName(), $items)) {
+                        $userItems[$childItem->getName()] = $childItem;
```

Comment:
If check rules here, we prevent unnecessary calls `getAllChildren`.

Reply:
https://github.com/yiisoft/rbac/pull/203/commits/6c6c1ee7aeb4aaaf023983d042055cb57d04e339